File: web/src/api/endpoints/providers.ts
import { createLogger } from '@/lib/logger'
import { getCsrfToken } from '@/utils/csrf'
import { IS_DEV_AUTH_BYPASS } from '@/utils/dev'
import { apiClient, unwrap, unwrapVoid } from '../client'
import type { ApiResponse } from '../types/http'
import type {
  AddAllowlistEntryRequest,
  CreateFromPresetRequest,
  CreateProviderRequest,
  DiscoverModelsResponse,
  DiscoveryPolicyResponse,
  LocalModelParams,
  ProbePresetResponse,
  ProviderConfig,
  ProviderHealthSummary,
  ProviderModelResponse,
  ProviderPreset,
  PullModelRequest,
  PullProgressEvent,
  RemoveAllowlistEntryRequest,
  TestConnectionRequest,
  TestConnectionResponse,
  UpdateModelConfigRequest,
  UpdateProviderRequest,
} from '../types/providers'

const log = createLogger('providers-api')

export async function listProviders(): Promise<Record<string, ProviderConfig>> {
  const response = await apiClient.get<ApiResponse<Record<string, ProviderConfig>>>('/providers')
  const raw = unwrap<Record<string, ProviderConfig>>(response)
  const result: Record<string, ProviderConfig> = Object.create(null) as Record<string, ProviderConfig>
  for (const [key, provider] of Object.entries(raw)) {
    if (key === '__proto__' || key === 'constructor' || key === 'prototype') continue
    result[key] = provider
  }
  return result
}

export async function getProvider(name: string): Promise<ProviderConfig> {
  const response = await apiClient.get<ApiResponse<ProviderConfig>>(`/providers/${encodeURIComponent(name)}`)
  return unwrap(response)
}

export async function getProviderModels(name: string): Promise<ProviderModelResponse[]> {
  const response = await apiClient.get<ApiResponse<ProviderModelResponse[]>>(`/providers/${encodeURIComponent(name)}/models`)
  return unwrap(response)
}

export async function getProviderHealth(name: string): Promise<ProviderHealthSummary> {
  const response = await apiClient.get<ApiResponse<ProviderHealthSummary>>(`/providers/${encodeURIComponent(name)}/health`)
  return unwrap(response)
}

export async function createProvider(data: CreateProviderRequest): Promise<ProviderConfig> {
  const response = await apiClient.post<ApiResponse<ProviderConfig>>('/providers', data)
  return unwrap(response)
}

export async function updateProvider(name: string, data: UpdateProviderRequest): Promise<ProviderConfig> {
  const response = await apiClient.put<ApiResponse<ProviderConfig>>(`/providers/${encodeURIComponent(name)}`, data)
  return unwrap(response)
}

export async function deleteProvider(name: string): Promise<void> {
  const response = await apiClient.delete<ApiResponse<null>>(`/providers/${encodeURIComponent(name)}`)
  unwrapVoid(response)
}

export async function testConnection(name: string, data?: TestConnectionRequest): Promise<TestConnectionResponse> {
  // Extended timeout: local providers (Ollama) may need to load models into memory
  const response = await apiClient.post<ApiResponse<TestConnectionResponse>>(
    `/providers/${encodeURIComponent(name)}/test`,
    data ?? {},
    { timeout: 120_000 },
  )
  return unwrap(response)
}

export async function listPresets(): Promise<ProviderPreset[]> {
  const response = await apiClient.get<ApiResponse<ProviderPreset[]>>('/providers/presets')
  return unwrap(response)
}

export async function createFromPreset(data: CreateFromPresetRequest): Promise<ProviderConfig> {
  const response = await apiClient.post<ApiResponse<ProviderConfig>>('/providers/from-preset', data)
  return unwrap(response)
}

export async function probePreset(presetName: string): Promise<ProbePresetResponse> {
  const response = await apiClient.post<ApiResponse<ProbePresetResponse>>('/providers/probe-preset', {
    preset_name: presetName,
  })
  return unwrap(response)
}

export async function discoverModels(
  name: string,
  presetHint?: string,
): Promise<DiscoverModelsResponse> {
  const params = presetHint ? { preset_hint: presetHint } : undefined
  const response = await apiClient.post<ApiResponse<DiscoverModelsResponse>>(
    `/providers/${encodeURIComponent(name)}/discover-models`,
    undefined,
    { params },
  )
  return unwrap(response)
}

export async function getDiscoveryPolicy(): Promise<DiscoveryPolicyResponse> {
  const response = await apiClient.get<ApiResponse<DiscoveryPolicyResponse>>('/providers/discovery-policy')
  return unwrap(response)
}

export async function addAllowlistEntry(data: AddAllowlistEntryRequest): Promise<DiscoveryPolicyResponse> {
  const response = await apiClient.post<ApiResponse<DiscoveryPolicyResponse>>('/providers/discovery-policy/entries', data)
  return unwrap(response)
}

export async function removeAllowlistEntry(data: RemoveAllowlistEntryRequest): Promise<DiscoveryPolicyResponse> {
  const response = await apiClient.post<ApiResponse<DiscoveryPolicyResponse>>('/providers/discovery-policy/remove-entry', data)
  return unwrap(response)
}

/** Encode a model ID for use in URL paths, preserving `/` for :path params. */
function encodeModelIdPath(modelId: string): string {
  return modelId.split('/').map(encodeURIComponent).join('/')
}

/** Process buffered SSE lines, dispatching events to the callback. */
function processSseLines(
  lines: string[],
  state: { currentEvent: string },
  onProgress: (event: PullProgressEvent) => void,
): void {
  for (const line of lines) {
    if (line.startsWith('event: ')) {
      state.currentEvent = line.slice(7).trim()
    } else if (line.startsWith('data: ')) {
      try {
        const parsed = JSON.parse(line.slice(6)) as PullProgressEvent
        if (state.currentEvent === 'error' || parsed.error) {
          const errorMsg = parsed.error ?? parsed.status ?? 'Pull failed'
          state.currentEvent = ''
          onProgress(parsed)
          throw new Error(errorMsg)
        }
        onProgress(parsed)
      } catch (err) {
        state.currentEvent = ''
        if (err instanceof SyntaxError) {
          log.warn('Malformed JSON in pull stream line')
          continue
        }
        if (err instanceof Error) throw err
      }
      state.currentEvent = ''
    }
  }
}

/**
 * Pull a model on a local provider via SSE streaming.
 *
 * Uses fetch + ReadableStream because the endpoint is POST-based
 * and EventSource only supports GET.
 */
export async function pullModel(
  name: string,
  modelName: string,
  onProgress: (event: PullProgressEvent) => void,
  signal?: AbortSignal,
): Promise<void> {
  const baseUrl = apiClient.defaults.baseURL ?? ''
  const url = `${baseUrl}/providers/${encodeURIComponent(name)}/models/pull`

  const csrfToken = getCsrfToken()
  const response = await fetch(url, {
    method: 'POST',
    credentials: 'include',
    headers: {
      'Content-Type': 'application/json',
      ...(csrfToken ? { 'X-CSRF-Token': csrfToken } : {}),
    },
    body: JSON.stringify({ model_name: modelName } satisfies PullModelRequest),
    signal,
  })

  if (!response.ok || !response.body) {
    if (response.status === 401 && !IS_DEV_AUTH_BYPASS) {
      // Server clears the session cookie. Sync Zustand auth state.
      import('@/stores/auth').then(({ useAuthStore }) => {
        useAuthStore.getState().handleUnauthorized()
      }).catch((importErr: unknown) => {
        log.error('Auth store cleanup failed during SSE 401 handling:', importErr)
        if (window.location.pathname !== '/login' && window.location.pathname !== '/setup') {
          window.location.href = '/login'
        }
      })
    }
    throw new Error(`Pull failed: HTTP ${response.status}`)
  }

  const reader = response.body.getReader()
  const decoder = new TextDecoder()
  let buffer = ''
  let receivedDone = false
  const sseState = { currentEvent: '' }

  while (true) {
    const { done, value } = await reader.read()
    if (done) break
    buffer += decoder.decode(value, { stream: true })

    const lines = buffer.split('\n')
    buffer = lines.pop() ?? ''

    processSseLines(lines, sseState, (event) => {
      if (event.done) receivedDone = true
      onProgress(event)
    })
  }

  // Process any remaining data in the buffer after the stream ends
  buffer += decoder.decode()
  if (buffer.trim()) {
    const trailing = buffer.split('\n')
    processSseLines(trailing, sseState, (event) => {
      if (event.done) receivedDone = true
      onProgress(event)
    })
  }

  if (!receivedDone) {
    throw new Error('Pull stream ended without completion event')
  }
}

export async function deleteModel(name: string, modelId: string): Promise<void> {
  const response = await apiClient.delete<ApiResponse<null>>(
    `/providers/${encodeURIComponent(name)}/models/${encodeModelIdPath(modelId)}`,
  )
  unwrapVoid(response)
}

export async function updateModelConfig(
  name: string,
  modelId: string,
  params: LocalModelParams,
): Promise<ProviderModelResponse> {
  const payload: UpdateModelConfigRequest = { local_params: params }
  const response = await apiClient.put<ApiResponse<ProviderModelResponse>>(
    `/providers/${encodeURIComponent(name)}/models/${encodeModelIdPath(modelId)}/config`,
    payload,
  )
  return unwrap(response)
}

File: src/synthorg/api/controllers/providers.py
"""Provider controller -- CRUD, connection testing, and presets."""

import asyncio
import json as _json
from collections.abc import (
    Mapping,  # noqa: TC003  # Litestar inspects runtime return-type annotation
)
from datetime import UTC, datetime, timedelta
from typing import TYPE_CHECKING, Annotated

if TYPE_CHECKING:
    from collections.abc import AsyncIterator

from litestar import Controller, delete, get, post, put
from litestar.datastructures import State  # noqa: TC002
from litestar.params import Parameter
from litestar.response import ServerSentEvent
from litestar.status_codes import HTTP_204_NO_CONTENT

from synthorg.api.dto import (
    ApiResponse,
    CreateFromPresetRequest,
    CreateProviderRequest,
    DiscoverModelsResponse,
    ProbePresetRequest,
    ProbePresetResponse,
    ProviderResponse,
    TestConnectionResponse,
    UpdateProviderRequest,
    to_provider_response,
)
from synthorg.api.dto import (
    TestConnectionRequest as ConnTestRequest,
)
from synthorg.api.dto_discovery import (
    AddAllowlistEntryRequest,
    DiscoveryPolicyResponse,
    RemoveAllowlistEntryRequest,
)
from synthorg.api.dto_providers import (
    ProviderModelResponse,
    PullModelRequest,
    UpdateModelConfigRequest,
    to_provider_model_response,
)
from synthorg.api.errors import (
    ApiError,
    ApiValidationError,
    ConflictError,
    NotFoundError,
)
from synthorg.api.guards import require_ceo_or_manager, require_read_access
from synthorg.api.path_params import PathName  # noqa: TC001
from synthorg.api.state import AppState  # noqa: TC001
from synthorg.observability import get_logger
from synthorg.observability.events.api import (
    API_MODEL_CAPABILITIES_LOOKUP_FAILED,
    API_MODEL_OPERATION_FAILED,
    API_PROVIDER_HEALTH_QUERIED,
    API_PROVIDER_USAGE_ENRICHMENT_FAILED,
    API_RESOURCE_CONFLICT,
    API_RESOURCE_NOT_FOUND,
    API_SSE_PULL_MODEL_FAILED,
    API_VALIDATION_FAILED,
)
from synthorg.providers.errors import (
    ProviderAlreadyExistsError,
    ProviderNotFoundError,
    ProviderValidationError,
)
from synthorg.providers.health import ProviderHealthSummary  # noqa: TC001
from synthorg.providers.presets import ProviderPreset, get_preset, list_presets
from synthorg.providers.probing import probe_preset_urls

logger = get_logger(__name__)


def _sse_error(msg: str) -> dict[str, object]:
    """Build a PullProgressEvent-shaped error dict for SSE."""
    return {
        "status": msg,
        "progress_percent": None,
        "total_bytes": None,
        "completed_bytes": None,
        "error": msg,
        "done": True,
    }


async def _enrich_with_usage(
    summary: ProviderHealthSummary,
    app_state: AppState,
    name: str,
) -> ProviderHealthSummary:
    """Enrich a health summary with token/cost data from CostTracker.

    Args:
        summary: Base health summary from the health tracker.
        app_state: Application state.
        name: Provider name.

    Returns:
        Enriched summary (or unchanged if enrichment is unavailable).
    """
    if not app_state.has_cost_tracker:
        return summary
    try:
        now = datetime.now(UTC)
        usage = await app_state.cost_tracker.get_provider_usage(
            name,
            start=now - timedelta(hours=24),
            end=now,
        )
        return summary.model_copy(
            update={
                "total_tokens_24h": usage.total_tokens,
                "total_cost_24h": usage.total_cost,
            },
        )
    except MemoryError, RecursionError:
        raise
    except Exception as exc:
        logger.warning(
            API_PROVIDER_USAGE_ENRICHMENT_FAILED,
            provider=name,
            error=str(exc),
            error_type=type(exc).__qualname__,
        )
        return summary


class ProviderController(Controller):
    """LLM provider management: CRUD, test, and presets."""

    path = "/providers"
    tags = ("providers",)

    # ── Read endpoints (read access) ─────────────────────────

    @get(
        "/presets",
        guards=[require_read_access],
    )
    async def get_presets(
        self,
        state: State,  # noqa: ARG002
    ) -> ApiResponse[tuple[ProviderPreset, ...]]:
        """List all available provider presets."""
        return ApiResponse(data=list_presets())

    @post(
        "/probe-preset",
        guards=[require_ceo_or_manager],
    )
    async def probe_preset(
        self,
        state: State,  # noqa: ARG002
        data: ProbePresetRequest,
    ) -> ApiResponse[ProbePresetResponse]:
        """Probe a preset's candidate URLs for reachability.

        Raises:
            ApiValidationError: If the preset is unknown.
        """
        preset = get_preset(data.preset_name)
        if preset is None:
            logger.warning(
                API_VALIDATION_FAILED,
                resource="preset",
                name=data.preset_name,
            )
            msg = f"Unknown preset: {data.preset_name!r}"
            raise ApiValidationError(msg)
        if not preset.candidate_urls:
            return ApiResponse(
                data=ProbePresetResponse(candidates_tried=0),
            )
        result = await probe_preset_urls(preset.name)
        return ApiResponse(
            data=ProbePresetResponse(
                url=result.url,
                model_count=result.model_count,
                candidates_tried=result.candidates_tried,
            ),
        )

    @get(
        "/",
        guards=[require_read_access],
    )
    async def list_providers(
        self,
        state: State,
    ) -> ApiResponse[Mapping[str, ProviderResponse]]:
        """List all configured providers (secrets stripped)."""
        app_state: AppState = state.app_state
        providers = await app_state.config_resolver.get_provider_configs()
        safe = {name: to_provider_response(p) for name, p in providers.items()}
        return ApiResponse(data=safe)

    @get(
        "/{name:str}",
        guards=[require_read_access],
    )
    async def get_provider(
        self,
        state: State,
        name: PathName,
    ) -> ApiResponse[ProviderResponse]:
        """Get a provider by name (secrets stripped).

        Raises:
            NotFoundError: If the provider is not found.
        """
        app_state: AppState = state.app_state
        providers = await app_state.config_resolver.get_provider_configs()
        provider = providers.get(name)
        if provider is None:
            msg = f"Provider {name!r} not found"
            logger.warning(API_RESOURCE_NOT_FOUND, resource="provider", name=name)
            raise NotFoundError(msg)
        return ApiResponse(data=to_provider_response(provider))

    @get(
        "/{name:str}/models",
        guards=[require_read_access],
    )
    async def list_models(
        self,
        state: State,
        name: PathName,
    ) -> ApiResponse[tuple[ProviderModelResponse, ...]]:
        """List models for a provider with runtime capabilities.

        Raises:
            NotFoundError: If the provider is not found.
        """
        app_state: AppState = state.app_state
        providers = await app_state.config_resolver.get_provider_configs()
        provider = providers.get(name)
        if provider is None:
            msg = f"Provider {name!r} not found"
            logger.warning(API_RESOURCE_NOT_FOUND, resource="provider", name=name)
            raise NotFoundError(msg)

        driver = None
        if app_state.has_provider_registry and name in app_state.provider_registry:
            driver = app_state.provider_registry.get(name)

        results: list[ProviderModelResponse] = []
        for model_config in provider.models:
            caps = None
            if driver is not None:
                try:
                    caps = await driver.get_model_capabilities(model_config.id)
                except MemoryError, RecursionError:
                    raise
                except Exception as exc:
                    logger.warning(
                        API_MODEL_CAPABILITIES_LOOKUP_FAILED,
                        provider=name,
                        model=model_config.id,
                        error=str(exc),
                        error_type=type(exc).__qualname__,
                    )
            results.append(to_provider_model_response(model_config, caps))
        return ApiResponse(data=tuple(results))

    @get(
        "/{name:str}/health",
        guards=[require_read_access],
    )
    async def get_provider_health(
        self,
        state: State,
        name: PathName,
    ) -> ApiResponse[ProviderHealthSummary]:
        """Get provider health summary (enriched with cost data).

        Raises:
            NotFoundError: If the provider is not found.
        """
        app_state: AppState = state.app_state
        providers = await app_state.config_resolver.get_provider_configs()
        if name not in providers:
            msg = f"Provider {name!r} not found"
            logger.warning(API_RESOURCE_NOT_FOUND, resource="provider", name=name)
            raise NotFoundError(msg)
        summary = await app_state.provider_health_tracker.get_summary(name)
        summary = await _enrich_with_usage(summary, app_state, name)
        logger.debug(
            API_PROVIDER_HEALTH_QUERIED,
            provider=name,
            health_status=summary.health_status.value,
            calls_24h=summary.calls_last_24h,
        )
        return ApiResponse(data=summary)

    # ── Write endpoints (write access) ───────────────────────

    @post(
        "/",
        guards=[require_ceo_or_manager],
    )
    async def create_provider(
        self,
        state: State,
        data: CreateProviderRequest,
    ) -> ApiResponse[ProviderResponse]:
        """Create a new provider.

        Args:
            state: Application state.
            data: Create provider request.

        Returns:
            Created provider response.

        Raises:
            ConflictError: If a provider with this name already exists.
            ApiValidationError: If the provider configuration fails
                validation.
        """
        app_state: AppState = state.app_state
        # Strip preset_name from external requests -- only
        # create_from_preset may set it (capability flag injection).
        safe_data = data.model_copy(update={"preset_name": None})
        try:
            config = await app_state.provider_management.create_provider(
                safe_data,
            )
        except ProviderAlreadyExistsError as exc:
            logger.warning(
                API_RESOURCE_CONFLICT,
                resource="provider",
                name=data.name,
            )
            raise ConflictError(str(exc)) from exc
        except ProviderValidationError as exc:
            logger.warning(
                API_VALIDATION_FAILED,
                resource="provider",
                error=str(exc),
            )
            raise ApiValidationError(str(exc)) from exc
        return ApiResponse(data=to_provider_response(config))

    @post(
        "/from-preset",
        guards=[require_ceo_or_manager],
    )
    async def create_from_preset(
        self,
        state: State,
        data: CreateFromPresetRequest,
    ) -> ApiResponse[ProviderResponse]:
        """Create a provider from a preset.

        Args:
            state: Application state.
            data: Preset-based creation request.

        Returns:
            Created provider response.

        Raises:
            ConflictError: If a provider with this name already exists.
            ApiValidationError: If the preset is unknown or config
                validation fails.
        """
        app_state: AppState = state.app_state
        try:
            config = await app_state.provider_management.create_from_preset(
                data,
            )
        except ProviderAlreadyExistsError as exc:
            logger.warning(
                API_RESOURCE_CONFLICT,
                resource="provider",
                name=data.name,
            )
            raise ConflictError(str(exc)) from exc
        except ProviderValidationError as exc:
            logger.warning(
                API_VALIDATION_FAILED,
                resource="provider",
                error=str(exc),
            )
            raise ApiValidationError(str(exc)) from exc
        return ApiResponse(data=to_provider_response(config))

    @put(
        "/{name:str}",
        guards=[require_ceo_or_manager],
    )
    async def update_provider(
        self,
        state: State,
        name: PathName,
        data: UpdateProviderRequest,
    ) -> ApiResponse[ProviderResponse]:
        """Update an existing provider.

        Args:
            state: Application state.
            name: Provider name.
            data: Partial update request.

        Returns:
            Updated provider response.

        Raises:
            NotFoundError: If the provider does not exist.
            ApiValidationError: If the update fails validation.
        """
        app_state: AppState = state.app_state
        try:
            config = await app_state.provider_management.update_provider(
                name,
                data,
            )
        except ProviderNotFoundError as exc:
            logger.warning(
                API_RESOURCE_NOT_FOUND,
                resource="provider",
                name=name,
            )
            raise NotFoundError(str(exc)) from exc
        except ProviderValidationError as exc:
            logger.warning(
                API_VALIDATION_FAILED,
                resource="provider",
                error=str(exc),
            )
            raise ApiValidationError(str(exc)) from exc
        return ApiResponse(data=to_provider_response(config))

    @delete(
        "/{name:str}",
        guards=[require_ceo_or_manager],
        status_code=HTTP_204_NO_CONTENT,
    )
    async def delete_provider(
        self,
        state: State,
        name: PathName,
    ) -> None:
        """Delete a provider.

        Args:
            state: Application state.
            name: Provider name.

        Raises:
            NotFoundError: If the provider does not exist.
        """
        app_state: AppState = state.app_state
        try:
            await app_state.provider_management.delete_provider(name)
        except ProviderNotFoundError as exc:
            logger.warning(
                API_RESOURCE_NOT_FOUND,
                resource="provider",
                name=name,
            )
            raise NotFoundError(str(exc)) from exc

    @post(
        "/{name:str}/discover-models",
        guards=[require_ceo_or_manager],
    )
    async def discover_models(
        self,
        state: State,
        name: PathName,
        preset_hint: Annotated[str, Parameter(max_length=64)] | None = None,
    ) -> ApiResponse[DiscoverModelsResponse]:
        """Discover available models from a provider endpoint.

        Queries the provider's API for available models and updates
        the provider configuration with any discovered models.  When
        ``base_url`` is not configured, returns an empty result.

        Args:
            state: Application state.
            name: Provider name.
            preset_hint: Optional preset name to guide endpoint
                selection (e.g. ``"ollama"``).

        Returns:
            Discovery result with found models.

        Raises:
            NotFoundError: If the provider does not exist.
        """
        app_state: AppState = state.app_state
        mgmt = app_state.provider_management
        try:
            discovered = await mgmt.discover_models_for_provider(
                name,
                preset_hint=preset_hint,
            )
        except ProviderNotFoundError as exc:
            logger.warning(
                API_RESOURCE_NOT_FOUND,
                resource="provider",
                name=name,
            )
            raise NotFoundError(str(exc)) from exc
        return ApiResponse(
            data=DiscoverModelsResponse(
                discovered_models=discovered,
                provider_name=name,
            ),
        )

    @post(
        "/{name:str}/test",
        guards=[require_ceo_or_manager],
    )
    async def test_connection(
        self,
        state: State,
        name: PathName,
        data: ConnTestRequest,
    ) -> ApiResponse[TestConnectionResponse]:
        """Test connectivity to a provider.

        Args:
            state: Application state.
            name: Provider name.
            data: Test connection request (includes optional model selection).

        Returns:
            Connection test result.

        Raises:
            NotFoundError: If the provider does not exist.
        """
        app_state: AppState = state.app_state
        try:
            result = await app_state.provider_management.test_connection(
                name,
                data,
            )
        except ProviderNotFoundError as exc:
            logger.warning(
                API_RESOURCE_NOT_FOUND,
                resource="provider",
                name=name,
            )
            raise NotFoundError(str(exc)) from exc
        return ApiResponse(data=result)

    # ── Discovery allowlist (read + write access) ──────────────

    @get(
        "/discovery-policy",
        guards=[require_read_access],
    )
    async def get_discovery_policy(
        self,
        state: State,
    ) -> ApiResponse[DiscoveryPolicyResponse]:
        """Return the current provider discovery SSRF allowlist.

        Args:
            state: Application state.

        Returns:
            Current discovery policy envelope.
        """
        app_state: AppState = state.app_state
        policy = await app_state.provider_management.get_discovery_policy()
        return ApiResponse(
            data=DiscoveryPolicyResponse.model_validate(
                policy,
                from_attributes=True,
            ),
        )

    @post(
        "/discovery-policy/entries",
        guards=[require_ceo_or_manager],
    )
    async def add_allowlist_entry(
        self,
        state: State,
        data: AddAllowlistEntryRequest,
    ) -> ApiResponse[DiscoveryPolicyResponse]:
        """Add a custom host:port entry to the discovery allowlist.

        Args:
            state: Application state.
            data: Request with the host:port entry to add.

        Returns:
            Updated discovery policy envelope.
        """
        app_state: AppState = state.app_state
        policy = await app_state.provider_management.add_custom_allowlist_entry(
            data.host_port,
        )
        return ApiResponse(
            data=DiscoveryPolicyResponse.model_validate(
                policy,
                from_attributes=True,
            ),
        )

    @post(
        "/discovery-policy/remove-entry",
        guards=[require_ceo_or_manager],
    )
    async def remove_allowlist_entry(
        self,
        state: State,
        data: RemoveAllowlistEntryRequest,
    ) -> ApiResponse[DiscoveryPolicyResponse]:
        """Remove a host:port entry from the discovery allowlist.

        Args:
            state: Application state.
            data: Request with the host:port entry to remove.

        Returns:
            Updated discovery policy envelope.
        """
        app_state: AppState = state.app_state
        policy = await app_state.provider_management.remove_custom_allowlist_entry(
            data.host_port,
        )
        return ApiResponse(
            data=DiscoveryPolicyResponse.model_validate(
                policy,
                from_attributes=True,
            ),
        )

    # ── Local model management ───────────────────────────────

    @post(
        "/{name:str}/models/pull",
        guards=[require_ceo_or_manager],
        media_type="text/event-stream",
    )
    async def pull_model(
        self,
        state: State,
        name: PathName,
        data: PullModelRequest,
    ) -> ServerSentEvent:
        """Pull a model on a local provider (SSE streaming).

        Args:
            state: Application state.
            name: Provider name.
            data: Pull request with model name.

        Returns:
            SSE stream of pull progress events.
        """
        app_state: AppState = state.app_state
        svc = app_state.provider_management

        async def _event_stream() -> AsyncIterator[dict[str, str]]:
            try:
                async for event in svc.pull_model(name, data.model_name):
                    if event.done and event.error:
                        event_type = "error"
                    elif event.done:
                        event_type = "complete"
                    else:
                        event_type = "progress"
                    yield {
                        "event": event_type,
                        "data": _json.dumps(event.model_dump()),
                    }
            except ProviderNotFoundError:
                yield {
                    "event": "error",
                    "data": _json.dumps(
                        _sse_error(f"Provider {name!r} not found"),
                    ),
                }
            except ProviderValidationError as exc:
                yield {
                    "event": "error",
                    "data": _json.dumps(_sse_error(str(exc))),
                }
            except asyncio.CancelledError:
                raise
            except MemoryError, RecursionError:
                raise
            except Exception as exc:
                logger.exception(
                    API_SSE_PULL_MODEL_FAILED,
                    provider=name,
                    model=data.model_name,
                    error=str(exc),
                    error_type=type(exc).__qualname__,
                )
                yield {
                    "event": "error",
                    "data": _json.dumps(
                        _sse_error(
                            f"Internal error: {type(exc).__name__}",
                        ),
                    ),
                }

        return ServerSentEvent(content=_event_stream())

    @delete(
        "/{name:str}/models/{model_id:path}",
        guards=[require_ceo_or_manager],
        status_code=HTTP_204_NO_CONTENT,
    )
    async def delete_model(
        self,
        state: State,
        name: PathName,
        model_id: Annotated[str, Parameter(max_length=256, min_length=1)],
    ) -> None:
        """Delete a model from a local provider.

        Args:
            state: Application state.
            name: Provider name.
            model_id: Model identifier (may contain colons).
        """
        app_state: AppState = state.app_state
        try:
            await app_state.provider_management.delete_model(name, model_id)
        except ProviderNotFoundError as exc:
            msg = f"Provider {name!r} not found"
            logger.warning(
                API_RESOURCE_NOT_FOUND,
                resource="provider",
                name=name,
            )
            raise NotFoundError(msg) from exc
        except ProviderValidationError as exc:
            logger.warning(
                API_VALIDATION_FAILED,
                resource="provider",
                name=name,
                error=str(exc),
            )
            raise ApiValidationError(str(exc)) from exc
        except ValueError as exc:
            logger.warning(
                API_RESOURCE_NOT_FOUND,
                resource="model",
                name=model_id,
                provider=name,
            )
            raise NotFoundError(str(exc)) from exc
        except RuntimeError as exc:
            logger.exception(
                API_MODEL_OPERATION_FAILED,
                resource="model",
                operation="delete",
                name=model_id,
                provider=name,
                error=str(exc),
            )
            raise ApiError(str(exc)) from exc

    @put(
        "/{name:str}/models/{model_id:path}/config",
        guards=[require_ceo_or_manager],
    )
    async def update_model_config(
        self,
        state: State,
        name: PathName,
        model_id: Annotated[str, Parameter(max_length=256, min_length=1)],
        data: UpdateModelConfigRequest,
    ) -> ApiResponse[ProviderModelResponse]:
        """Update per-model launch parameters for a local provider.

        Args:
            state: Application state.
            name: Provider name.
            model_id: Model identifier.
            data: New launch parameters.

        Returns:
            Updated model response.
        """
        app_state: AppState = state.app_state
        try:
            updated = await app_state.provider_management.update_model_config(
                name,
                model_id,
                data.local_params,
            )
        except ProviderNotFoundError as exc:
            msg = f"Provider {name!r} not found"
            logger.warning(
                API_RESOURCE_NOT_FOUND,
                resource="provider",
                name=name,
            )
            raise NotFoundError(msg) from exc
        except ProviderValidationError as exc:
            exc_msg = str(exc)
            if "not found" in exc_msg:
                logger.warning(
                    API_RESOURCE_NOT_FOUND,
                    resource="model",
                    name=model_id,
                    provider=name,
                )
                raise NotFoundError(exc_msg) from exc
            logger.warning(
                API_VALIDATION_FAILED,
                resource="provider",
                name=name,
                model=model_id,
                error=exc_msg,
            )
            raise ApiValidationError(exc_msg) from exc
        model = next(
            (m for m in updated.models if m.id == model_id),
            None,
        )
        if model is None:
            msg = f"Model {model_id!r} missing from updated config"
            logger.error(
                API_MODEL_OPERATION_FAILED,
                resource="model",
                operation="config_update",
                name=model_id,
                provider=name,
                error=msg,
            )
            raise ApiError(msg)
        return ApiResponse(data=to_provider_model_response(model))
